Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp

```diff
@@ -578,6 +578,32 @@ void FillRectWithGradient::dump(TextStream& ts, OptionSet<AsTextFlag>) const
     ts.dumpProperty("rect", rect());
 }
 
+FillRectWithGradientAndSpaceTransform::FillRectWithGradientAndSpaceTransform(const FloatRect& rect, Gradient& gradient, const AffineTransform& gradientSpaceTransform)
+    : m_rect(rect)
+    , m_gradient(gradient)
+    , m_gradientSpaceTransform(gradientSpaceTransform)
+{
+}
+
+FillRectWithGradientAndSpaceTransform::FillRectWithGradientAndSpaceTransform(FloatRect&& rect, Ref<Gradient>&& gradient, AffineTransform&& gradientSpaceTransform)
```

Comment:
Maybe you can have only this constructor and you make the RecorderImpl pass

```
append(FillRectWithGradientAndSpaceTransform(rect, Ref { gradient }, gradientSpaceTransform));
```

Reply:
Makes sense to check. I followed the existing pattern where most classes offer two constructors, e.g.

```
DrawDisplayListItems::DrawDisplayListItems(const Vector<Item>& items, const FloatPoint& destination)
    : m_items(items)
    , m_destination(destination)
{
}

DrawDisplayListItems::DrawDisplayListItems(Vector<Item>&& items, const FloatPoint& destination)
    : m_items(WTFMove(items))
    , m_destination(destination)
{
}
```

However, we can try to unify them, and get rid of the FIXME in the DisplayListItems header as well for both FillRectWithGradient and the new FillRectWithGradientAndSpaceTransform.

Reply:
One possibility is to keep only the constructor taking the rvalue references, e.g.:

```
diff --git a/Source/WebCore/platform/graphics/displaylists/DisplayListItems.h b/Source/WebCore/platform/graphics/displaylists/DisplayListItems.h
index 0e4218629b6c..5befb335467b 100644
--- a/Source/WebCore/platform/graphics/displaylists/DisplayListItems.h
+++ b/Source/WebCore/platform/graphics/displaylists/DisplayListItems.h
@@ -936,8 +936,7 @@ class FillRectWithGradient {
 public:
     static constexpr char name[] = "fill-rect-with-gradient";
 
-    WEBCORE_EXPORT FillRectWithGradient(const FloatRect&, Gradient&);
-    WEBCORE_EXPORT FillRectWithGradient(FloatRect&&, Ref<Gradient>&&);
+    WEBCORE_EXPORT FillRectWithGradient(const FloatRect&&, Ref<Gradient>&&);
 
     const FloatRect& rect() const { return m_rect; }
     const Ref<Gradient>& gradient() const { return m_gradient; }
```

Note the rather unfamiiar "const Foo&&" usage, that is required to pass on the "const FloatRect&" from the recorder as-is:

```
diff --git a/Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp b/Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp
index 69f9b7705378..061981b04ffd 100644
--- a/Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp
+++ b/Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp
@@ -274,12 +274,12 @@ void RecorderImpl::recordFillRectWithColor(const FloatRect& rect, const Color& c
 
 void RecorderImpl::recordFillRectWithGradient(const FloatRect& rect, Gradient& gradient)
 {
-    append(FillRectWithGradient(rect, gradient));
+    append(FillRectWithGradient(WTFMove(rect), Ref { gradient }));
 }
```

I will try this further and see if I hit any issue.

Reply:
DisplayListTests (API tests) require a change, that makes the code slightly harder to read:

```
diff --git a/Tools/TestWebKitAPI/Tests/WebCore/DisplayListTests.cpp b/Tools/TestWebKitAPI/Tests/WebCore/DisplayListTests.cpp
index d45c209e3951..200639129af4 100644
--- a/Tools/TestWebKitAPI/Tests/WebCore/DisplayListTests.cpp
+++ b/Tools/TestWebKitAPI/Tests/WebCore/DisplayListTests.cpp
@@ -74,7 +74,10 @@ TEST(DisplayListTests, AppendItems)
     for (int i = 0; i < 50; ++i) {
         list.append(SetInlineStroke(1.5));
         list.append(FillPath(path));
-        list.append(FillRectWithGradient(FloatRect { 1., 1., 10., 10. }, gradient));
+        {
+            auto rect = FloatRect { 1., 1., 10., 10. };
+            list.append(FillRectWithGradient(WTFMove(rect), Ref { gradient }));
+        }
         list.append(SetInlineFillColor(Color::red));
 #if ENABLE(INLINE_PATH_DATA)
         list.append(StrokeLine(PathDataLine { { 0., 0. }, { 10., 15. } }));
```

Anything else benefits so, I think we should still go with your suggestion, and get away with a single constructor.